### PR TITLE
Part of #2151 - Remove [Activate] from ViewComponents

### PR DESCRIPTION
**Pull request by rynowak**
This change removes [Activate] from ViewComponents. Accessing context
should be done through [ViewComponentContext]. Accessing services should
be done though constructor injection.

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentActivator.cs [hunk 1]
```diff
@@ -3,40 +3,38 @@
...
 using System.Reflection;
-using Microsoft.AspNet.Mvc.Rendering;
-using Microsoft.Framework.DependencyInjection;
 using Microsoft.Framework.Internal;
 
 namespace Microsoft.AspNet.Mvc.ViewComponents
 {
     /// <summary>
-    /// Represents the <see cref="IViewComponentActivator"/> that is registered by default.
+    /// A default implemented of <see cref="IViewComponentActivator"/>.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentActivator.cs]: implemented -> implementation

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs [hunk 2]
```diff
@@ -96,64 +99,95 @@ public ModelStateDictionary ModelState
...
+
+        [ViewComponentContext]
+        public ViewComponentContext ViewComponentContext
         {
             get
             {
-                if (_viewContext == null)
+                if (_viewComponentContext == null)
                 {
-                    // This should run only for the ViewComponent unit test scenarios
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs]: Why is this comment now incorrect?

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs [hunk 3]
```diff
@@ -96,64 +99,95 @@ public ModelStateDictionary ModelState
...
-                _viewContext = value;
+                _viewComponentContext = value;
             }
         }
 
         /// <summary>
-        /// Gets or sets the <see cref="ViewDataDictionary"/>.
+        /// Gets the <see cref="ViewContext"/>.
         /// </summary>
         [Activate]
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs]: Remove this line

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs [hunk 4]
```diff
@@ -96,64 +99,95 @@ public ModelStateDictionary ModelState
...
+        public ICompositeViewEngine ViewEngine
+        {
+            get
+            {
+                if (_viewEngine == null)
                 {
-                    // This should run only for the ViewComponent unit test scenarios
-                    _viewData = new ViewDataDictionary(ViewContext.ViewData);
+                    // May be null in unit-testing scenarios.
+                    var services = ViewComponentContext.ViewContext?.HttpContext?.RequestServices;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs]: Isn't it `ViewComponentContext` (not its `ViewContext` property) that needs the `null` check?  Could use `ViewContext` property instead.

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentActivator.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentActivator.cs]: Suggest mentioning `[ViewComponentContext]` and fact activator handles only `ViewComponentContext` properties with that attribute.

Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentActivator.cs]: implemented -> implementation

File: test/Microsoft.AspNet.Mvc.Core.Test/ViewComponentTests.cs [hunk 5]
```diff
@@ -180,7 +164,7 @@ public void ViewComponent_ViewContext_ViewData_ReturnsDefaultInstanceIfNull()
             Assert.Empty(viewComponent.ViewContext.ViewData);
             Assert.NotNull(viewComponent.ViewData);
             Assert.Empty(viewComponent.ViewData);
-            Assert.NotSame(viewComponent.ViewData, viewComponent.ViewContext.ViewData);
+            Assert.Same(viewComponent.ViewData, viewComponent.ViewContext.ViewData);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ViewComponentTests.cs]: Glad to here this inconsistency go bye-bye

**Comment by dougbu** [comment]
:watch: mostly because I feel I'm missing something

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentActivator.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentActivator.cs]: :+1: to all

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs]: No idea. Put it back.

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs]: fixt

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponent.cs]: ViewComponentContext is guaranteed to not be null. Removed some extra ?s from other places in code.

File: test/Microsoft.AspNet.Mvc.Core.Test/ViewComponentTests.cs (hunk 5 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ViewComponentTests.cs]: It's back!

It's not an inconsistency, it's a defensive copy

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ViewComponentTests.cs]: We need to discuss this. The "defensive copy" causes a user experience different from that in other contexts e.g. within a controller action.

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponentContextAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNet.Mvc.ViewComponents
+{
+    [AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true)]
+    public class ViewComponentContextAttribute : Attribute
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponentContextAttribute.cs]: Doc this one.

**Comment by rynowak** [comment]
Updated - changed to defensive copy of `ViewContext` instead of a copy of just `ViewData`

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponentContext.cs [hunk 7]
```diff
@@ -29,32 +42,54 @@ public class ViewComponentContext
         {
             ViewComponentDescriptor = viewComponentDescriptor;
             Arguments = arguments;
-            ViewContext = viewContext;
-            Writer = writer;
+
+            // We want to create a defensive copy of the VDD here so that changes done in the VC
+            // aren't visible in the calling view.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/ViewComponentContext.cs]: Nice. This behaviour is also consistent with tag helper activations, HTML helper recursions, and so on.

File: test/Microsoft.AspNet.Mvc.Core.Test/ViewComponentTests.cs (hunk 5 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ViewComponentTests.cs]: It's back to being changed :dancers:

File: test/WebSites/ActivatorWebSite/Components/ValueComponent.cs [hunk 8]
```diff
@@ -1,19 +0,0 @@
...
-
-using Microsoft.AspNet.Mvc;
-
-namespace ActivatorWebSite
-{
-    [ViewComponent(Name = "Value")]
-    public class ValueComponent : ViewComponent
-    {
-        [Activate]
-        public ViewService MyViewService { get; set; }
```
Comment (dougbu) [inline test/WebSites/ActivatorWebSite/Components/ValueComponent.cs]: Do we have tests elsewhere that test constructor injection into a view component. Fine to file a follow-up bug if not. Idea is to cover the general scenario of using some service or other in a view component.

Comment (rynowak) [inline test/WebSites/ActivatorWebSite/Components/ValueComponent.cs]: Yes. There's a preponderance of tests for this aspect of activation

Comment (dougbu) [inline test/WebSites/ActivatorWebSite/Components/ValueComponent.cs]: Nevermind. Just a couple of files further down...

**Comment by dougbu** [comment]
:shipit:

**Pull request closed without merging**